### PR TITLE
[STORM-3684] use real compId instead of constant _system for receive-queue V2 metrics

### PR DESCRIPTION
**Pull request by RuiLi8080**
## What is the purpose of the change

When registering receive-queue V2 metrics for each executor, we should use the its correspondent component Id instead of constant system component Id.

## How was the change tested
Launch a local cluster with a WordCount topology. Enable `org.apache.storm.metrics2.reporters.ConsoleStormReporter` to output metrics to worker log. We will see the long name as :
`storm.worker.<topoName>.<hostname>.<compId>.<taskId>.<port>.<metric-name>`
The compId part now reflects the correct component for the taskId.
```
2020-08-04 14:22:44.911 c.c.m.ConsoleReporter metrics-console-reporter-1-thread-1 [INFO] storm.worker.wc1-1-1596568923.192_168_0_19.count.10.6700-receive-queue-pct_full
2020-08-04 14:22:44.911 c.c.m.ConsoleReporter metrics-console-reporter-1-thread-1 [INFO]              value = 0.0
2020-08-04 14:22:44.912 c.c.m.ConsoleReporter metrics-console-reporter-1-thread-1 [INFO] storm.worker.wc1-1-1596568923.192_168_0_19.count.10.6700-receive-queue-population
2020-08-04 14:22:44.912 c.c.m.ConsoleReporter metrics-console-reporter-1-thread-1 [INFO]              value = 0
2020-08-04 14:22:44.912 c.c.m.ConsoleReporter metrics-console-reporter-1-thread-1 [INFO] storm.worker.wc1-1-1596568923.192_168_0_19.count.10.6700-receive-queue-sojourn_time_ms
2020-08-04 14:22:44.912 c.c.m.ConsoleReporter metrics-console-reporter-1-thread-1 [INFO]              value = 0.0
```

### CODE REVIEW

File: storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java [hunk 1]
```diff
@@ -717,11 +718,20 @@ private Assignment getLocalAssignment(IStormClusterState stormClusterState, Stri
 
         IWaitStrategy backPressureWaitStrategy = IWaitStrategy.createBackPressureWaitStrategy(topologyConf);
         Map<List<Long>, JCQueue> receiveQueueMap = new HashMap<>();
+
+        Map<Integer, String> taskIdToCompId = StormCommon.stormTaskInfo(topology, topologyConf);
```
Comment (Ethanlm) [inline storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java:722]: Is it possible to reuse `this.taskToComponent`?

File: storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java [hunk 2]
```diff
@@ -199,7 +200,6 @@ public WorkerState(Map<String, Object> conf,
         }
         Collections.sort(localTaskIds);
         this.topologyConf = topologyConf;
-        this.topology = ConfigUtils.readSupervisorTopology(conf, topologyId, AdvancedFSOps.make(conf));
```
Comment (Ethanlm) [inline storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java:202]: anything specific to this change?

File: storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java (hunk 1 above)
Comment (RuiLi8080) [reply storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java]: Yes, updated. Thanks for pointing out.

File: storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java (hunk 2 above)
Comment (RuiLi8080) [reply storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java:202]: I actually updated. Basically mkReceiveQueueMap() function needs to know `this.taskToComponent` which relies on `this.topology`. So we need to move them up.

File: storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java [hunk 3]
```diff
@@ -186,6 +186,9 @@ public WorkerState(Map<String, Object> conf,
         this.isWorkerActive = new CountDownLatch(1);
         this.isTopologyActive = new AtomicBoolean(false);
         this.stormComponentToDebug = new AtomicReference<>();
+        this.topology = ConfigUtils.readSupervisorTopology(conf, topologyId, AdvancedFSOps.make(conf));
+        this.taskToComponent = StormCommon.stormTaskInfo(topology, topologyConf);
+        // mkReceiveQueueMap relies on taskToComponent which relies on topology above
```
Comment (Ethanlm) [inline storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java:191]: I would change `mkReceiveQueueMap` method to include `taskToComponent` parameter to make this relationship more obvious

File: storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java [hunk 4]
```diff
@@ -717,11 +719,19 @@ private Assignment getLocalAssignment(IStormClusterState stormClusterState, Stri
 
         IWaitStrategy backPressureWaitStrategy = IWaitStrategy.createBackPressureWaitStrategy(topologyConf);
         Map<List<Long>, JCQueue> receiveQueueMap = new HashMap<>();
+
         for (List<Long> executor : executors) {
             List<Integer> taskIds = StormCommon.executorIdToTasks(executor);
+            int taskId = taskIds.get(0);
+            String compId;
+            if (taskId == -1) {
```
Comment (Ethanlm) [inline storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java:727]: I would  use `SYSTEM_TASK_ID` instead of `-1`

Comment (RuiLi8080) [reply storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java]: Yes, updated

File: storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java (hunk 3 above)
Comment (RuiLi8080) [reply storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java]: SUre

File: storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java [hunk 5]
```diff
@@ -717,11 +718,19 @@ private Assignment getLocalAssignment(IStormClusterState stormClusterState, Stri
 
         IWaitStrategy backPressureWaitStrategy = IWaitStrategy.createBackPressureWaitStrategy(topologyConf);
         Map<List<Long>, JCQueue> receiveQueueMap = new HashMap<>();
+
         for (List<Long> executor : executors) {
             List<Integer> taskIds = StormCommon.executorIdToTasks(executor);
+            int taskId = taskIds.get(0);
+            String compId;
+            if (taskId == Constants.SYSTEM_TASK_ID) {
+                compId = Constants.SYSTEM_COMPONENT_ID;
```
Comment (agresch) [inline storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java:727]: Should StormCommon.stormTaskInfo() be addressing this mapping?

Comment (RuiLi8080) [reply storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java:727]: StormCommon.stormTaskInfoImpl() address taskID starting from 1. 
WorkerState.readWorkerExecutors() additionally add Constants.SYSTEM_EXECUTOR_ID (-1) before reading from assignment. This system executor is not present in `taskToComponent` map.

File: storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java [hunk 6]
```diff
@@ -704,7 +704,8 @@ private Assignment getLocalAssignment(IStormClusterState stormClusterState, Stri
         }
     }
 
-    private Map<List<Long>, JCQueue> mkReceiveQueueMap(Map<String, Object> topologyConf, Set<List<Long>> executors) {
+    private Map<List<Long>, JCQueue> mkReceiveQueueMap(Map<String, Object> topologyConf,
+                                                       Set<List<Long>> executors,  Map<Integer, String> taskToComponent) {
```
Comment (Ethanlm) [inline storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java:708]: nit: space

Comment (RuiLi8080) [reply storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java]: updated

**Comment by Ethanlm** [comment]
travis failure seems unrelated.

**Pull request merged**